### PR TITLE
Extend foreach command by --unique and --recursive

### PR DESCRIPTION
**Pull request by t-wissmann**
This adds two flags to foreach: --unique removes duplicates (or more
precisely, aliases) of objects and --recursive recurses into subobjects.
To ensure termination --recursive implicitly also uses --unique. The
recursion is done in a breadth-first way such that deeply nested objects
are visited later.

The main application for --unique is to avoid the duplicate
clients.focus and the main application of --recursive is to traverse all
frames in the tree layout.

### CODE REVIEW

File: tests/test_meta_commands.py [hunk 1]
```diff
@@ -746,12 +746,108 @@ def test_foreach_object_completion(hlwm):
...
+    hlwm.call('split bottom')
+    hlwm.call('split left')
+
+    def object_paths():
+        return hlwm.call(['foreach', 'X', '', '--recursive', 'echo', 'X']) \
+                   .stdout.splitlines()
+
+    without_client = object_paths()
+
+    # and put the a client in the frame leaf:
```
Comment (blarz) [inline tests/test_meta_commands.py:782]: typo the a?

File: tests/test_meta_commands.py [hunk 2]
```diff
@@ -746,12 +746,108 @@ def test_foreach_object_completion(hlwm):
...
+
+    found_in_without_client = False
+    for tag_id in ['0', 'focus']:
+        frame_object = f'tags.{tag_id}.tiling.focused_frame'
+        assert frame_object not in with_client
+        if frame_object in without_client:
+            found_in_without_client = True
+
+    assert found_in_without_client, \
+        "The frame must be listed under tags. before client creation"
```
Comment (blarz) [inline tests/test_meta_commands.py:799]: replace . with ,?

**Review by blarz — COMMENTED** [review]
Minor issues, otherwise looks good to me!

File: tests/test_meta_commands.py (hunk 1 above)
Comment (t-wissmann) [reply tests/test_meta_commands.py]: I've removed 'a'

File: tests/test_meta_commands.py (hunk 2 above)
Comment (t-wissmann) [reply tests/test_meta_commands.py]: actually, I meant the object `tags.`, I've simply replaced it with `'tags'`

**Comment by t-wissmann** [comment]
Thanks for checking, I've fixed both! :-)

File: tests/test_meta_commands.py (hunk 2 above)
Comment (blarz) [reply tests/test_meta_commands.py]: Aaah, thanks for clearing that up! :)

**Comment by blarz** [comment]
> Thanks for checking, I've fixed both! :-)

Thanks :)

**Review by blarz — APPROVED** [review]

**Pull request merged**